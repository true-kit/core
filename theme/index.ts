export {
	getTheme,
	ThemeProvider,
	createThemeFor,
	composeThemes,
} from './theme';

export {
	Theme,
} from './theme.types';