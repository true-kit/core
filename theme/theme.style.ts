export class ThemeStyle {
	state = {};

	constructor(public name: string, public classes: object, private classNames: Function) {
		// console.log(name, classes, classNames.toString());
	}

	set(name: string, state: string | boolean) {
		this.state[name] = state;
		return this;
	}

	toString() {
		return this.classNames();
	}
}

ThemeStyle.prototype.constructor = String;