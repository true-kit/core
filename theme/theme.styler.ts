export class ThemeStyler {
	state = {};

	constructor(public name: string, public classes: object, private classNames: Function) {
		// console.log(name, classes, classNames.toString());
	}

	set(name: string, state: string | boolean) {
		this.state[name] = state;
		return this;
	}

	toDOMProps<T>(props: React.AllHTMLAttributes<T> = {}): React.AllHTMLAttributes<T> {
		props.className = (this as any) as string;
		return props;
	}

	toString() {
		return this.classNames();
	}
}

ThemeStyler.prototype.constructor = String;