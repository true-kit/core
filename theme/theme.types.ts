import { IRuleDefinitions } from '@artifact-project/css';

export type LikeComponent<T extends Theme<any>> = (props: {theme?: T}) => any;
export type GetComponentTheme<C> = C extends (props: {theme?: infer T}) => any ? T : never;

export type ThemeModsSpec = {
	[name:string]: string | boolean | undefined;
}

export type ThemeElementsSpec = {
	[name:string]: boolean | ThemeModsSpec;
}

export type ThemeSpec = {
	host: ThemeModsSpec;
	elements: ThemeElementsSpec;
}

export type Theme<TS extends ThemeSpec> = {
	readonly spec: {
		host: Required<TS['host']>;
		elements: Required<TS['elements']>;
	};
	readonly Owner: LikeComponent<Theme<TS>>;
	readonly cssRules: IRuleDefinitions;
	readonly registry: object | null;

	readonly classes: {
		host: ThemeClasses<TS['host']>;
		elements: ThemeElementsClasses<TS['elements']>
	}

	// Get theme from `host` or element
	for<
		K extends keyof TS['elements'],
		N extends 'host' | K,
	>(name: N): N extends K
		? string & ThemeElement<N, TS['elements'][K]>
		: string & ThemeElement<N, TS['host']>
	;

	withRegistry(registry: object): Theme<TS>;
}

export type ThemeClasses<T extends ThemeModsSpec> = {$root: string} & {
	[K in keyof T]: string;
};

export type ThemeElementsClasses<T extends ThemeElementsSpec> = {
	[K in keyof T]?: T[K] extends ThemeModsSpec ? ThemeClasses<T[K]> : {$root: string};
};

export type ThemeElement<N, T extends boolean | ThemeModsSpec> = {
	name: N;
	toString(): string;
} & (T extends ThemeModsSpec
	? {
		set<K extends keyof T>(name: K, state: T[K]): ThemeElement<N, T>;
		classes: {$root: string} & {[K in keyof T]: string};
	}
	: {
		set(): ThemeElement<N, T>;
		classes: {$root: string};
	}
);

export type CSSProps = Partial<React.CSSProperties>;

// Rule
export type ThemeRule<
	M extends ThemeModsSpec | ThemeElementsSpec,
	E extends ThemeElementsSpec,
	A extends ThemeModsSpec | ThemeElementsSpec = E, // Adjacent Sibling Selector
> = CSSProps
	& ThemePseudoRules<M, E>
	& ThemeNotRule<M, E, A>
	& ThemeSelfRefRule<M, E, A>
	& {
	'+'?: ThemeRule<M, E, A>
	':nested'?: ThemeAnyRule & ThemeElementsRules<E> & ThemeNotRule<M, E>;
}

// '*' — AnyRule
export type ThemeAnyRule = {
	'*'?: {
		[selector:string]: CSSProps;
	};
};

// ':not(...)'
export type ThemeNotRule<
	M extends ThemeModsSpec | ThemeElementsSpec,
	E extends ThemeElementsSpec,
	A extends ThemeModsSpec | ThemeElementsSpec = E,
> = {
	':not'?: ThemeSelfRefRule<M, E, A> & ThemePseudoRules<M, E> & {
		[K in keyof M]?: M[K] extends string
			? {[V in M[K]]?: ThemeRule<M, E>}
			: (M[K] extends ThemeModsSpec
				? ThemeRule<M[K], E> & ThemeModifiersRule<M[K], E>
				: ThemeRule<M, E>
			)
	};
}

// '&' — Self reference rule
export type ThemeSelfRefRule<
	M extends ThemeModsSpec | ThemeElementsSpec,
	E extends ThemeElementsSpec,
	A extends ThemeModsSpec | ThemeElementsSpec = E,
> = A extends ThemeModsSpec
	? ThemeAnyRule & ThemeNotRule<M, E> & {
		'&'?: (M extends ThemeElementsSpec
			? ThemeRule<M, E, A> & ThemeModifiersRule<A, E>
			: ThemeHostRule<A, E>
		);
	}

	: ThemeElementsRules<E> & ThemeAnyRule & ThemeNotRule<M, E> & {
		'&'?: (M extends ThemeModsSpec
			? ThemeRule<A, E> & ThemeModifiersRule<M, E>
			: ThemeRule<A, E>
		);
	}
;

// ':pseudo' — Pseudo selector rule
export type ThemePseudoRules<
	M extends ThemeModsSpec | ThemeElementsSpec,
	E extends ThemeElementsSpec,
> = {
	':first-child'?: ThemeRule<M, E>;
	':last-child'?: ThemeRule<M, E>;
	':before'?: ThemeRule<M, E>;
	':after'?: ThemeRule<M, E>;
	':hover'?: ThemeRule<M, E>;
	':active'?: ThemeRule<M, E>;
	':visited'?:ThemeRule<M, E>;
	':focus'?: ThemeRule<M, E>;
}

// Rules for 'host' & 'elements'
export type ThemeRules<T extends ThemeSpec> = {
	host: ThemeHostRule<T['host'], T['elements']>;
	elements: ThemeElementsRules<T['elements']>;
}

// ':host' — Host Rule
export type ThemeHostRule<
	M extends ThemeModsSpec, // Modifiers
	E extends ThemeElementsSpec, // Elements
> = ThemeRule<M, E, M> & ThemeModifiersRuleWithSelf<M, E>;

// Modifiers Rule
export type ThemeModifiersRule<
	M extends ThemeModsSpec, // Modifiers
	E extends ThemeElementsSpec,
> = {
	':modifiers'?: {
		[K in keyof M]?: M[K] extends string
			? {[V in M[K]]?: ThemeRule<M, E>}
			: ThemeRule<M, E>
	};
}

export type ThemeModifiersRuleWithSelf<
	M extends ThemeModsSpec, // Modifiers
	E extends ThemeElementsSpec,
> = {
	':modifiers'?: {
		[K in keyof M]?: M[K] extends string
			? {[V in M[K]]?: ThemeRule<M, E>} & {':self'?: ThemeRule<M, E>}
			: ThemeRule<M, E>
	};
}

export type ThemeElementsRules<E extends ThemeElementsSpec> = {
	[K in keyof E]?: E[K] extends ThemeModsSpec
		? ThemeRule<E, E> & ThemeModifiersRule<E[K], E>
		: ThemeRule<E, E>
	;
}

export type ThemeRegistry = {
	map: Map<LikeComponent<any>, Theme<any>> | null;
	overrides: ThemeOverrideIndex | null;
}

export type ThemeOverride = {
	theme: Theme<any>;
	xpath: LikeComponent<any>[];
}

export type ThemeOverrideIndex = Map<LikeComponent<any>, ThemeOverride>;
