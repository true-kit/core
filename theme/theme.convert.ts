import { ThemeRules, ThemeRule } from './theme.types';
import { IRuleDefinitions, CSSMap, css } from '@artifact-project/css';

export function createToStringCode(code: string[], name: string, classes: object) {
	if (typeof classes === 'string') {
		code.push(`
			if (n === "${name}") {
				return (c || '');
			}
		`);
		return;
	}

	const str = Object.entries(classes).map(([key, val]) => {
		if (key === '$root') {
			return name === 'host' ? `(c.$root || '')` : '';
		} else if (typeof val === 'object') {
			return `(k = '${key}', s[k] ? (c[k].$root ? ' ' + c[k].$root : '') + ' ' + (c[k][s[k]] || '') : '')`;
		} else {
			return `(k = '${key}', s[k] ? ' ' + (c[k] || '') : '')`;
		}
	}).filter(x => !!x).join(' + ');

	code.push(`
		if (n === "${name}") {
			return (${str ? str : '""'});
		}
	`);
}

const SEP = '--';
const COLON_CODE = ':'.charCodeAt(0);

type SelectorToken = {
	type: '+' | '.' | 'fn' | ':';
	raw: string;
}

class Selector {
	constructor(raw: SelectorToken['raw'], type: SelectorToken['type'] = '.', private tokens: SelectorToken[] = []) {
		this.tokens.push({raw, type});
	}

	add(raw: SelectorToken['raw'], type: SelectorToken['type'] = '.') {
		return new Selector(raw, type, this.tokens.slice(0));
	}

	toString() {
		let fnOpened = false;
		let prevType: SelectorToken['type'];

		return this.tokens.reduce((s, {raw, type}, idx) => {
			if (type === '.') {
				if (raw === '&') {
					s += '.' + this.tokens[0].raw;
				} else if (prevType !== '.' && idx > 0) {
					s += '.' + this.tokens[0].raw + SEP + raw;
				} else {
					s += (prevType === '.' ? SEP : '.') + raw;
				}
			} else if (type === ':') {
				s += raw;
			} else {
				if (fnOpened) {
					s += ')';
					fnOpened = false;
				}

				if (type === '+') {
					s += ' + ';
				} else if (type === 'fn') {
					s += raw + '('
					fnOpened = true;
				}
			}

			if (fnOpened && (this.tokens.length - idx === 1)) {
				s += ')';
				fnOpened = false;
			}

			prevType = type;
			return s;
		}, '');
	}
}

export function convertThemeRulesToCSSRules(rules: ThemeRules<any>, css = {}) {
	Object.entries(rules).forEach(([name, rule]) => {
		convertThemeRuleToCSSRules(new Selector(name), rule, css);
	});

	return css as IRuleDefinitions;
}

function convertThemeRuleToCSSRules(sel: Selector, rule: ThemeRule<any, any>, css: object) {
	Object.entries(rule).forEach(([key, value]) => {
		if (value && typeof value === 'object') {
			if (key.charCodeAt(0) === COLON_CODE) {
				if (key === ':modifiers' || key === ':self') {
					convertThemeRuleToCSSRules(sel, value, css);
				} else if (key === ':not') {
					convertThemeRuleToCSSRules(sel.add(key, 'fn'), value, css);
				} else {
					convertThemeRuleToCSSRules(sel.add(key, ':'), value, css);
				}
			} else if (key === '+') {
				convertThemeRuleToCSSRules(sel.add('+', '+'), value, css);
			} else if (key === '&') {
				convertThemeRuleToCSSRules(sel.add('&'), value, css);
			} else {
				convertThemeRuleToCSSRules(sel.add(key), value, css);
			}
		} else {
			const name = sel.toString();
			css[name] = css[name] || {};
			css[name][key] = value;
		}
	});
}

export function convertCSSRulesToCSSClasses<T extends IRuleDefinitions>(rules: T): CSSMap<T> {
	return css(rules)
}

export function convertCSSClassesToThemeClasses<T extends IRuleDefinitions, M extends CSSMap<T>>(map: M) {
	const classes = {
	};

	Object.entries(map).forEach(([key, val]) => {
		const [elem, mod, state] = key.split(SEP);

		if (classes[elem] == null) {
			classes[elem] = map[elem] ? {$root: map[elem]} : {};
		}

		if (mod != null && !state && classes[elem][mod] == null) {
			classes[elem][mod] = val;
		}

		if (state != null) {
			if (!classes[elem][mod] || typeof classes[elem][mod] !== 'object') {
				classes[elem][mod] = classes[elem][mod] ? {$root: classes[elem][mod]} : {};
			}

			classes[elem][mod][state] = val;
		}
	});

	!classes.hasOwnProperty('host') && (classes['host'] = {});
	!classes.hasOwnProperty('elements') && (classes['elements'] = {});

	return classes;
}
