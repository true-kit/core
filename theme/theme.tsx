import {
	ThemeRules,
	ThemeSpec,
	Theme,
	ThemeRegistry,
	LikeComponent,
	ThemeOverride,
	GetThemeFromDescriptor,
	ThemeOverrideIndex,
	AllThemes,
} from './theme.types';

import {
	convertThemeRulesToCSSRules,
	convertCSSRulesToCSSClasses,
	convertCSSClassesToThemeClasses,
	createToStringCode,
} from './theme.convert';

import { ThemeStyle } from './theme.style';

import { DescriptorWithMeta, Last, Cast } from '../core.types';
import { EnvContextEntry } from '../env/env.types';
import { createEnvContextProvider, getEnvContext, getActiveEnvScope } from '../env/env';


export function createThemeRegistryFor<
	D extends DescriptorWithMeta<any, any>
>(
	descriptor: D,
	themesMap: AllThemes<D>,
	overrides?: ThemeOverride[] | null,
): ThemeRegistry {
	return createThemeRegistry(
		Object.values(themesMap).map(theme => theme) as Theme<any>[],
		overrides,
	);
}

export function createThemeFor<
	D extends DescriptorWithMeta<any, {theme?: Theme<any>}>,
	TS extends ThemeSpec = NonNullable<D['meta']['theme']>['spec'],
>(descriptor: D, rules: ThemeRules<TS>): Theme<TS> {
	const cssRules = convertThemeRulesToCSSRules(rules);
	const cssClasses = convertCSSRulesToCSSClasses(cssRules);
	const classes = convertCSSClassesToThemeClasses(cssClasses);
	const toStringCode = [
		`
		var k;
		var n = this.name;
		var s = this.state;
		var c = this.classes;
		`
	];

	createToStringCode(toStringCode, 'host', classes['host']);
	Object.entries(classes['elements']).forEach(([name, classes]) => {
		createToStringCode(toStringCode, name, classes);
	});

	const classNames = Function(`
		${toStringCode.join('\n')}
		return '';
	`);

	function _for(name: string) {
		const store = this.store;

		if (store === null || store.hasOwnProperty(name)) {
			const style = new ThemeStyle(
				name,
				name === 'host' ? classes[name] : classes['elements'][name],
				classNames,
			) as any;

			if (store === null) {
				return style
			}

			store[name] = style;
		}

		return store[name];
	}

	return {
		descriptor: descriptor as any,
		for: _for,
		cssRules,
		classes,
		store: null,
		persist(store: object) {
			const wr = Object.create(this);
			wr.store = store;
			return wr;
		},
	} as Theme<TS>;
}

const nullTheme = createThemeFor(null as any, {
	host: {},
	elements: {},
});

export const ThemeProvider = createEnvContextProvider('theme');

export function getTheme<
	D extends DescriptorWithMeta<any, {theme?: Theme<any>}>,
	T extends Theme<any> = NonNullable<D['meta']['theme']>,
>(descriptor: D, props: {theme?: T}): T {
	let theme = props.theme;

	if (theme == null) {
		theme = nullTheme as T;

		let ctx = getEnvContext();

		while (ctx && ctx.theme) {
			const {
				map,
				overrides,
			} = ctx.theme;

			if (
				overrides !== null
				&& getActiveEnvScope() !== null
				&& overrides.has(descriptor)
			) {
				const overTheme = getThemeOverride(overrides.get(descriptor)!, ctx);
				if (overTheme !== null) {
					theme = overTheme as T;
					break;
				}
			}

			if (map !== null && map.has(descriptor)) {
				theme = map.get(descriptor) as T;
				break;
			}

			ctx = ctx.parent!;
		}
	}

	return theme;
}

function getThemeOverride(list: ThemeOverride[], ctx: EnvContextEntry): Theme<any> | null {
	for (let i = 0; i < list.length; i++) {
		let {value: theme, xpath} = list[i];
		let scope = getActiveEnvScope();

		XPATH: for (let x = 0, xn = xpath.length; x < xn; x++) {
			const descr = xpath[x];

			while (true) {
				scope = scope!.parent;

				if (scope === null || scope.ctx === ctx) {
					theme = null as any;
					break XPATH;
				}

				if (scope.owner === descr) {
					break;
				}
			}
		}

		if (theme !== null) {
			return theme;
		}
	}

	return null;
}

export function createThemeRegistry(
	themes: Theme<any>[] | null,
	overrides?: ThemeOverride[] | null,
): ThemeRegistry {
	return {
		map: !themes ? null : themes.reduce((map, theme) => {
			map.set(theme.descriptor, theme);
			return map;
		}, new Map),

		overrides: !overrides ? null : overrides.reduce((index, override) => {
			const xpath = override.xpath.slice();
			const Node = xpath.pop();

			if (xpath.length >= 1 && Node) {
				index.set(Node, (index.get(Node) || []).concat({
					value: override.value,
					xpath,
				}));
			} else {
				console.warn('[@truekit/core: theme] Invalid override: ', override);
			}

			return index;
		}, new Map as ThemeOverrideIndex),
	};
}

export function createThemeOverrideFor<
	X extends DescriptorWithMeta<any, {theme?: Theme<any>}>[]
>(...xpath: X): (theme: GetThemeFromDescriptor<Last<X>>) => ThemeOverride {
	return (theme): ThemeOverride => ({
		value: theme as any,
		xpath,
	});
}