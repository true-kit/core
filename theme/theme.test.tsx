import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {
	createThemeFor,
	getTheme,
	createThemeRegistry,
	createThemeOverrideFor,
	ThemeProvider,
	withThemeScopeEnv,
} from './theme';
import { resetCSS } from '@artifact-project/css';
import { Theme } from './theme.types';
// import * as classnames from 'classnames';
// import { now } from '@perf-tools/performance';

type TextProps = {
	children?: React.ReactNode;
	size?: 'big' | 'small';
	block?: boolean;
	type?: 'default' | 'error';
	disabled?: boolean;
	theme?: Theme<{
		host: Pick<TextProps, 'size' | 'disabled' | 'block' | 'type'> & {
			multiline: 'on' | 'off';
		};
		elements: {
			value: boolean;
		};
	}>;
};

type IconProps = {
	size?: 'big' | 'small';
	theme?: Theme<{
		host: Pick<IconProps, 'size'>;
		elements: {
		};
	}>;
};

function Text(props: TextProps) {
	return withThemeScopeEnv(Text, () => {
		const hostTheme = getTheme(Text, props).for('host');
		const result = <div className={hostTheme}>{props.children}</div>;

		props.size && hostTheme.set('size', props.size);
		props.disabled && hostTheme.set('disabled', true);

		return result;
	});
}

function Icon(props: IconProps) {
	return withThemeScopeEnv(Icon, () => {
		const hostTheme = getTheme(Icon, props).for('host');
		const result = <i className={hostTheme}/>;

		props.size && hostTheme.set('size', props.size);

		return result;
	});
}

beforeEach(() => {
	process.env.NODE_ENV = 'production';
	resetCSS(0);
});

describe('createThemeFor', () => {
	it('host', () => {
		const theme = createThemeFor(Text)({
			host: {color: '#333'},
			elements: {},
		});

		expect(theme.Owner).toBe(Text);
		expect(theme.cssRules).toEqual({
			'.host': {color: '#333'},
		});
		expect(theme.classes).toEqual({
			host: {$root: '_0'},
			elements: {},
		});
		expect(theme.for('host').toString()).toBe('_0');
	});

	it('for equal for', () => {
		const theme = createThemeFor(Text)({
			host: {color: '#333'},
			elements: {},
		});
		const host1 = theme.for('host');
		const host2 = theme.for('host');

		expect(host1).not.toBe(host2);
		expect(host1 + '').toBe(host2 + '');
	});

	it('withRegistry', () => {
		const theme = createThemeFor(Text)({
			host: {color: '#333'},
			elements: {},
		}).withRegistry({});
		const host1 = theme.for('host');
		const host2 = theme.for('host');

		expect(host1).toBe(host2);
		expect(host1 + '').toBe(host2 + '');
	});

	it('host - element', () => {
		const theme = createThemeFor(Text)({
			host: {color: '#333'},
			elements: {value: {color: 'red'}},
		});

		expect(theme.Owner).toBe(Text);
		expect(theme.cssRules).toEqual({
			'.host': {color: '#333'},
			'.elements--value': {color: 'red'},
		});
		expect(theme.classes).toEqual({
			host: {$root: '_0'},
			elements: {value: '_1'},
		});
		expect(theme.for('value').toString()).toBe('_1');
	});

	it('host - modifiers', () => {
		const theme = createThemeFor(Text)({
			host: {
				color: '#333',
				':modifiers': {
					size: {
						':self': {fontSize: '100%'},
						big: {fontSize: '200%'},
						small: {fontSize: '50%'},
					},
					disabled: {
						opacity: .5,
					},
				},
			},
			elements: {},
		});

		expect(theme.cssRules).toEqual({
			'.host': {color: '#333'},
			'.host--size': {fontSize: '100%'},
			'.host--size--big': {fontSize: '200%'},
			'.host--size--small': {fontSize: '50%'},
			'.host--disabled': {opacity: .5},
		});

		expect(theme.classes).toEqual({
			host: {
				$root: '_0',
				size: {$root: '_1', big: '_2', small: '_3'},
				disabled: '_4',
			},
			elements: {},
		});
		expect(theme.for('host').toString()).toBe('_0');
		expect(theme.for('host').set('size', 'small').toString()).toBe('_0 _1 _3');
		expect(theme.for('host').set('size', 'big').toString()).toBe('_0 _1 _2');
		expect(theme.for('host').set('disabled', true).toString()).toBe('_0 _4');
		expect(theme.for('host').set('size', 'small').set('disabled', true).toString()).toBe('_0 _1 _3 _4');
	});

	it('host + &', () => {
		const theme = createThemeFor(Text)({
			host: {
				color: '#333',
				'+': {
					'&': {
						marginLeft: 5,
						'+': {
							':not': {
								'&': {
									marginLeft: 10,
									'+': {'&': {marginLeft: 15}},
									':not': {
										':first-child': {
											marginLeft: 0,
										},
									},
								},
							},
						},
					},
				},
			},
			elements: {},
		});

		expect(theme.cssRules).toEqual({
			'.host': {color: '#333'},
			'.host + .host': {marginLeft: 5},
			'.host + .host + :not(.host)': {marginLeft: 10},
			'.host + .host + :not(.host) + .host': {marginLeft: 15},
			'.host + .host + :not(.host):not(:first-child)': {marginLeft: 0},
		});

		expect(theme.classes).toEqual({
			host: {$root: '_0'},
			elements: {},
		});
	});

	it('host - modifiers - not', () => {
		const theme = createThemeFor(Text)({
			host: {
				color: '#333',
				':modifiers': {
					block: {
						width: '100%',
						':not': {
							multiline: {on: {wordBreak: 'normal', whiteSpace: 'nowrap'}},
						},
					},
					multiline: {off: {wordBreak: 'normal', whiteSpace: 'nowrap'}},
					type: {
						':self': {fontWeight: 'normal'},
						default: {fontSize: '100%'},
						error: {color: 'red'},
					},
				},
			},
			elements: {},
		});

		expect(theme.cssRules).toEqual({
			'.host': {color: '#333'},
			'.host--block': {width: '100%'},
			'.host--block:not(.host--multiline--on)': {wordBreak: 'normal', whiteSpace: 'nowrap'},
			'.host--multiline--off': {wordBreak: 'normal', whiteSpace: 'nowrap'},
			'.host--type': {fontWeight: 'normal'},
			'.host--type--default': {fontSize: '100%'},
			'.host--type--error': {color: 'red'},
		});

		expect(theme.classes).toEqual({
			host: {
				$root: '_0',
				block: '_1',
				multiline: {
					on: '_2',
					off: '_3',
				},
				type: {
					$root: '_4',
					default: '_5',
					error: '_6',
				},
			},
			elements: {},
		});
	});
});

const root = document.createElement('div');

function render(fragment: JSX.Element) {
	ReactDOM.render(
		fragment,
		root,
	);
	return root.innerHTML;
}

describe('react', () => {
	it('inline theme', () => {
		const theme = createThemeFor(Text)({
			host: {
				color: 'red',
				':modifiers': {
					size: {small: {fontSize: '50%'}},
				},
			},
			elements: {},
		});

		expect(render(
			<Text theme={theme} size="small">Wow!</Text>,
		)).toMatchSnapshot();
	});

	it('without context', () => {
		expect(render(
			<Text>Wow!</Text>,
		)).toMatchSnapshot();
	});

	it('with context', () => {
		const rootTheme = createThemeRegistry([
			createThemeFor(Icon)({
				host: {
					display: 'inline-block',
					':modifiers': {
						size: {big: {fontSize: 300}},
					},
				},
				elements: {},
			}),
		]);

		const theme = createThemeRegistry([
			createThemeFor(Text)({
				host: {
					color: '#333',
					':modifiers': {
						disabled: {color: 'red'},
					},
				},
				elements: {},
			}),
		]);

		expect(render(
			<ThemeProvider value={rootTheme}>
				<ThemeProvider value={theme}>
					<Text disabled>Wow!</Text>
					<Icon size="big"/>
				</ThemeProvider>,
			</ThemeProvider>,
		)).toMatchSnapshot();
	});
});

describe('overrides', () => {
	it('createThemeOverrideFor', () => {
		const textTheme = createThemeFor(Text)({
			host: {color: '#333'},
			elements: {},
		});
		const iconTheme = createThemeFor(Icon)({
			host: {color: 'black'},
			elements: {},
		});
		const redIconTheme = createThemeFor(Icon)({
			host: {color: 'red'},
			elements: {},
		});
		const blueIconTheme = createThemeFor(Icon)({
			host: {color: 'blue'},
			elements: {},
		});
		const rootTheme = createThemeRegistry([textTheme, iconTheme], []);
		const redTheme = createThemeRegistry(null, [createThemeOverrideFor(Text, Icon)(redIconTheme)]);
		const blueTheme = createThemeRegistry(null, [createThemeOverrideFor(Text, Icon)(blueIconTheme)]);
		const icon = <Icon size="small"/>;

		expect(render(
			<ThemeProvider value={rootTheme}>
				<ThemeProvider value={redTheme}>
					<Text>{'\n'}
						{'\t'}red: <Icon size="small"/>{'\n'}
					</Text>{'\n'}

					<Text>{'\n'}
						<ThemeProvider value={blueTheme}>
							{'\t'}red: {icon}{'\n'}
							{'\t'}<Text>{'\n'}
								{'\t\t'}blue: {icon}{'\n'}
							{'\t'}</Text>
						</ThemeProvider>{'\n'}
					</Text>{'\n'}

					black: <Icon size="small"/>{'\n'}
				</ThemeProvider>

				black: <Icon size="small"/>
			</ThemeProvider>,
		)).toMatchSnapshot();
	});
});

// it('performance', () => {
// 	type TestProps = {
// 		theme?: Theme<{
// 			primary: boolean;
// 			warning: boolean;
// 			danger: boolean;
// 			fluid: boolean;
// 			large: boolean;
// 			flat: boolean;
// 			disabled: boolean;
// 			mobile: boolean;
// 			auto: boolean;
// 			icon: boolean;
// 		}>;
// 	};
// 	const cssRule = {color: 'inherit'};
// 	const TestCmp = (_: TestProps) => <div/>;
// 	const theme = createThemeFor(TestCmp)({
// 		':host': cssRule,
// 		primary: cssRule,
// 		warning: cssRule,
// 		danger: cssRule,
// 		fluid: cssRule,
// 		large: cssRule,
// 		flat: cssRule,
// 		disabled: cssRule,
// 		mobile: cssRule,
// 		auto: cssRule,
// 		icon: cssRule,
// 	});
// 	const maxIter = 1e3;
// 	const classes = {
// 			base: theme.classes[':host'],
// 			...theme.classes,
// 	};
// 	const states = Array.from({length: maxIter}).map(() => ({
// 			primary: Math.random() > .5,
// 			warning: Math.random() > .5,
// 			danger: Math.random() > .5,
// 			fluid: Math.random() > .5,
// 			large: Math.random() > .5,
// 			flat: Math.random() > .5,
// 			disabled: Math.random() > .5,
// 			mobile: Math.random() > .5,
// 			auto: Math.random() > .5,
// 			icon: Math.random() > .5,
// 	}));
// 	const pkgCSSTheme = [];
// 	const pkgClassNames = [];

// 	// pkg: classnames
// 	let pkgClassNamesTime = -now();
// 	for (let i = 0; i < maxIter; i++) {
// 		const s = states[i];
// 		const c = classes;
// 		pkgClassNames.push(classnames(
// 			c.base,
// 			{
// 				[c.primary]: s.primary,
// 				[c.warning]: s.warning,
// 				[c.danger]: s.danger,
// 				[c.fluid]: s.fluid,
// 				[c.large]: s.large,
// 				[c.flat]: s.flat,
// 				[c.disabled]: s.disabled,
// 				[c.mobile]: s.mobile,
// 				[c.auto]: s.auto,
// 				[c.icon]: s.icon,
// 			},
// 		));
// 	}
// 	pkgClassNamesTime += now();

// 	// pkg: css/theme
// 	let pkgCSSThemeTime = -now();
// 	for (let i = 0; i < maxIter; i++) {
// 		const s = states[i];
// 		const style = theme.create();

// 		style.set('primary', s.primary);
// 		style.set('warning', s.warning);
// 		style.set('danger', s.danger);
// 		style.set('fluid', s.fluid);
// 		style.set('large', s.large);
// 		style.set('flat', s.flat);
// 		style.set('disabled', s.disabled);
// 		style.set('mobile', s.mobile);
// 		style.set('auto', s.auto);
// 		style.set('icon', s.icon);

// 		pkgCSSTheme.push(style.toString());
// 	}
// 	pkgCSSThemeTime += now();

// 	expect(pkgCSSTheme).toEqual(pkgClassNames);
// 	expect(pkgCSSThemeTime).toBeLessThan(pkgClassNamesTime);
// });