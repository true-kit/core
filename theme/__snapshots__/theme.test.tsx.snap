// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overrides createThemeOverrideFor 1`] = `
"<div class=\\"_0\\">
	red: <i class=\\"_2\\"></i>
</div>
<div class=\\"_0\\">
	red: <i class=\\"_2\\"></i>
	<div class=\\"_0\\">
		blue: <i class=\\"_3\\"></i>
	</div>
	<div class=\\"_0\\">
		purple: <i class=\\"_4\\"></i>
	</div>
</div>
black: <i class=\\"_1\\"></i>
black: <i class=\\"_1\\"></i>"
`;

exports[`react createThemeRegistryFor 1`] = `"<button class=\\"_1\\"><i class=\\"_0\\"></i><div class=\\"_2\\">ok</div></button>"`;

exports[`react inline theme 1`] = `"<div class=\\"_0 _1\\">Wow!</div>"`;

exports[`react with context 1`] = `"<div class=\\"_2 _3\\">Wow!</div><i class=\\"_0 _1\\"></i>,"`;

exports[`react without (context & inline) 1`] = `"<div class=\\"\\">Wow!</div>"`;
